### ISSUE TITLE
Inaccuracy in timestamps

### ISSUE THREAD

**Issue by Nasreen123** [pnkfelix/tango#13]
- timestamps set for files are losing nansecond level precision
- timestamps seem to be getting muddled, with generated files being given an older time stamp

Steps to reproduce:

1) Delete main.md, the project should only have main.rs

2) Use tango to generate a new main.md file

3) Check the time stamps of the files (typing `ls -l --full-time` in the terminal to get them):
```
-rw-rw-r-- 1 nasreen nasreen 3369 2017-09-01 13:42:18.651217000 +0100 main.md
-rw-rw-r-- 1 nasreen nasreen 3357 2017-09-01 13:42:18.651217146 +0100 main.rs
-rw-rw-r-- 1 nasreen nasreen    0 2017-09-01 13:42:18.651217000 +0100 tango.stamp
```

   This is strange because it seems like main.rs is younger!

4) Delete main.md again, and regenerate it

5) Check the timestamps again, they are still:
```
-rw-rw-r-- 1 nasreen nasreen 3369 2017-09-01 13:42:18.651217000 +0100 main.md
-rw-rw-r-- 1 nasreen nasreen 3357 2017-09-01 13:42:18.651217146 +0100 main.rs
-rw-rw-r-- 1 nasreen nasreen    0 2017-09-01 13:42:18.651217000 +0100 tango.stamp
```
   This is strange because time has passed and they haven't changed

**Comment by pnkfelix** [comment]
Just for the public record (I already mentioned this directly to @Nasreen123 when we were chatting 1:1): it *does* seem strange at first that time has passed, and yet the generated files seem to have the same timestamps that they had from the original construction.

However, tango is playing tricks here, where it manually revises the timestamps of files it creates.

In particular (if my memory is correct), it *tries* to ensure two properties:

 1. Re-use the timestamp for each source file as the time stamp for the corresponding generated file. E.g. our goal is that `timestamp(main.md) == timestamp(main.rs)` after tango has a successful run.
 2. Use the youngest timestamp of all of the source files (which will likewise be the timestamp of all the generated files as well, assuming property 1 is actually enforced) as the timestamp for the `tango.stamp` file.

This way, we know that each file was created based on the content with the corresponding time stamp, and the `tango.stamp` file tells us the age of our youngest generated content (and thus we know we have to re-run tango if at some point in the future some source file ends up being even younger than that).

----

My current hypothesis is that somewhere in `timestamp.rs`, I messed something up and truncated the timestamps to millisecond precision rather than nanosecond precision. That, *or* the Rust API's for setting timestamps are doing such truncation under-the-hood.

There are two ways I can imagine to deal with this:

 1. Fix the bug, so that we set timestamps with the same nanosecond precision that the source files have, or
 2. Change the timestamps of the source files *as well as* the generated files. 

Strategy 2 will continue to enforce property 1 listed above. But it may have bad repercussions for interacting with other tools that do not expect timestamps for source files to go back in time, even by such minute amounts.

**Comment by pnkfelix** [comment]
Ah, I bet this is the cause for the imprecision (`tango` uses the `filetime` crate under the hood for manipulating the timestamps): https://github.com/alexcrichton/filetime/issues/9

**Comment by pnkfelix** [comment]
Oh, another strategy that I should have thought of when I made my list in the [earlier comment](https://github.com/pnkfelix/tango/issues/13#issuecomment-326931815):

 3. Instead of back-dating the source file (i.e. strategy 2, which I implemented in PR #14), I could instead have tango be looser in its own comparison, in that it could first truncate all inputs to millisecond precision when comparing the source and target timestamps. This would probably present fewer problems for compatibility with other tools...

**Comment by pnkfelix** [comment]
And further thought leads to the following refinements of strategies 2 and 3:

 2. (b) First re-stamp the target file. Then check the stamps on the source and target. Only attempt to re-stamp the source if the two now mismatch.

 3. (b) Before running tango at all, do a local test of the host system to see if one can record nanosecond level deviations in the timestamps. (E.g. read the `tango.stamp` timestamp, add 1 nanosecond to it, re-stamp the file with that incremented value, re-read the new stamp, and see if its different from the value we first read.)
     * If we *can* record nanosecond level precision, then use nanosecond granularity when comparing timestamps of files.
     * If we *can't* record nanosecond level precision, then use millisecond granularity when comparing timestamps of files.

At this point I've landed the original strategy 3 (and backed out my implementation of strategy 2). If the filetime crate issue (https://github.com/alexcrichton/filetime/issues/9) is fixed, then one could/should then take the time to implement strategy 3 b.
